File: scripts/gen_model_embeddings.py
#!/usr/bin/env python3
"""
Generate model-embeddings.json using the I3 Proxy embeddings API.
Source text is built from model-data.js: name + purpose + useCase (+ category/industry as context).

Usage:
  I3_API_KEY=... python3 scripts/gen_model_embeddings.py [--base http://localhost:8000]
"""

import os
import re
import sys
import json
import time
import argparse
from pathlib import Path
import urllib.request


def load_model_data(js_path: Path) -> dict:
    """Parse model-data.js and extract {name: {purpose, useCase, category, industry}}"""
    content = js_path.read_text(encoding="utf-8")
    # Regex similar to backend, tolerant of escaped quotes/newlines
    pattern = r'"([^"\\]+)":\s*\{\s*"purpose":\s*"([^"\\]*(?:\\.[^"\\]*)*)",\s*"useCase":\s*"([^"\\]*(?:\\.[^"\\]*)*)",\s*"category":\s*"([^"\\]*)",\s*"industry":\s*"([^"\\]*)"'
    models = {}
    for match in re.finditer(pattern, content, re.DOTALL):
        name = match.group(1)
        def clean(s: str) -> str:
            s = s.replace('\\"', '"').replace('\\n', ' ').replace('\n', ' ')
            # strip HTML tags if any
            s = re.sub(r'<[^>]+>', '', s)
            return s.strip()

        purpose = clean(match.group(2))
        use_case = clean(match.group(3))
        category = match.group(4).strip()
        industry = match.group(5).strip()
        models[name] = {
            "purpose": purpose,
            "useCase": use_case,
            "category": category,
            "industry": industry,
        }
    if not models:
        raise RuntimeError("No models parsed from model-data.js")
    return models


def build_card_text(name: str, d: dict) -> str:
    parts = [name]
    if d.get("purpose"):
        parts.append(d["purpose"]) 
    if d.get("useCase"):
        parts.append(d["useCase"]) 
    # Include category/industry to give more context for similarity
    if d.get("category"):
        parts.append(f"Category: {d['category']}")
    if d.get("industry"):
        parts.append(f"Industry: {d['industry']}")
    return "\n".join(parts)


def post_embeddings(base_url: str, api_key: str, texts: list) -> list:
    url = f"{base_url.rstrip('/')}/embeddings"
    payload = {
        "model": "i3-embedding",
        "input": texts,
    }
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(url, data=data, headers={
        "Content-Type": "application/json",
        "I3-API-Key": api_key,
    })
    with urllib.request.urlopen(req) as resp:
        body = resp.read().decode("utf-8")
    j = json.loads(body)
    # Expect proxy shape
    items = j.get("data", {}).get("data", [])
    return [item.get("embedding", []) for item in items]


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--base", default=os.environ.get("I3_PROXY_BASE", "http://localhost:8000"))
    args = parser.parse_args()

    api_key = os.environ.get("I3_API_KEY")
    if not api_key:
        print("Missing I3_API_KEY", file=sys.stderr)
        sys.exit(1)

    root = Path(__file__).resolve().parents[1]
    js_path = root / "model-data.js"
    out_path = root / "model-embeddings.json"

    models = load_model_data(js_path)
    names = list(models.keys())
    print(f"Parsed {len(names)} models from {js_path}")

    # Build texts
    items = [(name, build_card_text(name, models[name])) for name in names]

    out = []
    BATCH = 64
    for i in range(0, len(items), BATCH):
        batch = items[i:i+BATCH]
        texts = [t for _, t in batch]
        embs = post_embeddings(args.base, api_key, texts)
        if len(embs) != len(batch):
            raise RuntimeError(f"Embedding count mismatch at batch {i}: got {len(embs)} expected {len(batch)}")
        for (name, _), emb in zip(batch, embs):
            out.append({"name": name, "embedding": emb})
        print(f"Embedded {min(i+BATCH, len(items))} / {len(items)}")
        time.sleep(0.1)

    out_path.write_text(json.dumps(out), encoding="utf-8")
    print(f"Wrote {out_path} items: {len(out)}")


if __name__ == "__main__":
    main()

File: scripts/gen-model-embeddings.js
const fs = require('fs');
const path = require('path');
const fetch = require('node-fetch');

// Load MODEL_DATA from the browser bundle file by evaluating it in a sandbox-like scope.
// If direct import fails due to ESM/CJS differences, fallback to reading the file and extracting JSON.
function loadModelData() {
    const filePath = path.resolve(__dirname, '..', 'model-data.js');
    const code = fs.readFileSync(filePath, 'utf8');
    // Quick-and-safe extraction: evaluate in a Function with minimal scope
    const sandbox = { module: { exports: {} } };
    const wrapped = new Function('module', 'exports', 'window', code + '\n;return module.exports || {};');
    const mod = wrapped(sandbox.module, sandbox.module.exports, {});
    if (mod && mod.MODEL_DATA) return mod.MODEL_DATA;
    throw new Error('Unable to load MODEL_DATA from model-data.js');
}

function cardText(name, d) {
    return [name, d.purpose, d.useCase, `Category: ${d.category}`, `Industry: ${d.industry}`]
        .filter(Boolean).join('\n');
}

async function embedBatch(texts, model = 'i3-embedding') {
    const baseURL = process.env.I3_PROXY_BASE || 'http://localhost:8000';
    const apiKey = process.env.I3_API_KEY;
    if (!apiKey) throw new Error('Missing I3_API_KEY');
    const res = await fetch(`${baseURL}/embeddings`, {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
            'I3-API-Key': apiKey
        },
        body: JSON.stringify({ model, input: texts })
    });
    if (!res.ok) throw new Error(await res.text());
    const json = await res.json();
    // Proxy shape: { success, data: { data: [ { embedding } ] } }
    const arr = (json && json.data && Array.isArray(json.data.data)) ? json.data.data : [];
    return arr.map(x => x.embedding);
}

async function main() {
    // Require proxy API key
    if (!process.env.I3_API_KEY) {
        console.error('Missing I3_API_KEY');
        process.exit(1);
    }

    const MODEL_DATA = loadModelData();
    const entries = Object.entries(MODEL_DATA);
    const items = entries.map(([name, d]) => ({ name, text: cardText(name, d) }));

    const out = [];
    const B = 64;
    for (let i = 0; i < items.length; i += B) {
        const batch = items.slice(i, i + B);
        const embs = await embedBatch(batch.map(x => x.text));
        for (let j = 0; j < batch.length; j++) {
            out.push({ name: batch[j].name, embedding: embs[j] });
        }
        console.log(`Embedded ${Math.min(i + B, items.length)} / ${items.length}`);
        await new Promise(r => setTimeout(r, 150));
    }

    const outPath = path.resolve(process.cwd(), 'model-embeddings.json');
    fs.writeFileSync(outPath, JSON.stringify(out));
    console.log('Wrote', outPath, 'items:', out.length);
}

main().catch(e => { console.error(e); process.exit(1); });


